src/platform/intel/cavs/platform.c

```diff
@@ -543,7 +543,41 @@ void platform_wait_for_interrupt(int level)
 }
 #endif
 
+#define ADSP_IMR_MAGIC_VALUE		0x02468ACE
+#define IMR_L1_CACHE_ADDRESS		0xB0000000
+
+struct imr_header {
+	uint32_t adsp_imr_magic;
+	uint32_t structure_version;
+	uint32_t structure_size;
+	uint32_t imr_state;
+	uint32_t imr_size;
+	void *imr_restore_vector;
+};
+
+struct imr_state {
+	struct imr_header header;
+	uint8_t reserved[0x1000 - sizeof(struct imr_header)];
+};
+
+struct irm_layout {
+	uint8_t     css_reserved[0x1000];
+	struct imr_state    imr_state;
+};
+
+static void imr_layout_update(void *vector)
+{
+	struct irm_layout *imr_layout = (struct irm_layout *)(IMR_L1_CACHE_ADDRESS);
+
+	dcache_invalidate_region(imr_layout, sizeof(*imr_layout));
+	imr_layout->imr_state.header.adsp_imr_magic = ADSP_IMR_MAGIC_VALUE;
+	imr_layout->imr_state.header.imr_restore_vector = vector;
+	dcache_writeback_region(imr_layout, sizeof(*imr_layout));
```

Comment:
Agree with @lgirdwood, no idea what this does.

I guess this uses the cache to write everything back to DDR/IMR, but I don't see a platform_resume where everything is restored?

Shouldn't there be a platform_resume for consistency?

Reply:
> Agree with @lgirdwood, no idea what this does.
> 
> I guess this uses the cache to write everything back to DDR/IMR, but I don't see a platform_resume where everything is restored?

Yes, as IMR_L1_CACHE_ADDRESS is a cache address, we need to write it back to make sure it will take effect at the subsequent resuming.

> 
> Shouldn't there be a platform_resume for consistency?

The ROM code will read this from IMR at the subsequent run and decide (e.g. combine with checking if FW_PURGE IPC from host got) if it can use the previous IMR FW directly. So this here is only a host->FW->ROM one way configuration, no symmetric platform_resume() to clear the configuration needed.

Reply:
> The ROM code will read this from IMR at the subsequent run and decide (e.g. combine with checking if FW_PURGE IPC from host got) if it can use the previous IMR FW directly. So this here is only a host->FW->ROM one way configuration, no symmetric platform_resume() to clear the configuration needed.

ok, this info is needed in the comment.